src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs

```diff
@@ -164,7 +165,7 @@ public IReadOnlyList<PackageSpec> GetClosure(string rootUniqueName)
             }
 
             var projectsByUniqueName = _projects
-                .ToDictionary(t => t.Value.RestoreMetadata.ProjectUniqueName, t => t.Value, PathUtility.GetStringComparerBasedOnOS());
+                .ToDictionary(t => t.Value.RestoreMetadata.ProjectUniqueName, t => t.Value, _projects.Count, PathUtility.GetStringComparerBasedOnOS());
```

Comment:
Great find. I learnt few performance tips reviewing this PR. Thank you.

After reading comments in https://github.com/dotnet/runtime/issues/14366 and https://github.com/dotnet/runtime/issues/15246 issue, I navigated to below link. https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq/src/System/Linq/ToCollection.cs#L35-L74

It looks like `.ToDictionary` is optimized if the source is an `array` or `list`.

I took a step back and analyzed where the dictionary is used and how the items to added to the source `_projects` which is a sorted dictionary. `projectsUniqueName` is used as a key when adding items to the dictionary.

https://github.com/NuGet/NuGet.Client/blob/b102b7587f0286d83dad8ee357caded56bd0ab17/src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs#L216-L226

I removed the invocation to `.ToDictionary` in my https://github.com/NuGet/NuGet.Client/compare/dev-kartheep-ms-dependencygraph-perf?expand=1 private branch. I ran unit tests locally and found no issues. If CI is happy https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4928824&view=results then I am not sure if the new extension method is required now.

Reply:
Yep, removing the allocation completion is ever better! Feel free to submit yours, I'll close this one.

Reply:
CI passed. Created PR https://github.com/NuGet/NuGet.Client/pull/4133